WIP - Implementing AArch64 trampoline support

hsoontie opened the pull request:
WIP - Implementing AArch64 trampoline support

Signed-off-by: Hillary Soontiens <hsoontie@unb.ca>

hsoontie commented:
Here is a start on my code for AArch64 trampolines, which I am trying to base off of ARM's section in the same file. There isn't much here right now that's new, but I wanted to ask for some suggestions.

There were a few things I wanted to know before trying fill this out some more:
- How can I find ARM64 hex encodings equivalent to all the ones in the ARM functions, if needed? For example, how would I find the equivalent of `*buffer = 0xe51ff004;`?
- What do the "#if defined TR_HOST_<architecture>" parts in the ARM functions mean? Ex:
	```
	#if defined(TR_HOST_ARM)
	extern void armCodeSync(uint8_t *, uint32_t);
	#endif /* TR_HOST_ARM */
	```

0xdaryl commented on runtime/compiler/runtime/Trampoline.cpp:
```diff
@@ -858,6 +858,71 @@ void armCodeCacheParameters(int32_t *trampolineSize, void **callBacks, int32_t *
+#define TRAMPOLINE_SIZE         16 // ?
+
+#define BRANCH_FORWARD_LIMIT    0x01fffffc // What should these two be for AArch64?
+#define BRANCH_BACKWARD_LIMIT   0xfe000000
```
The right way of asking for this information is not via macros (the old way), but via the generalized target queries:
```
TR::Compiler->target.cpu.maxUnconditionalBranchImmediateForwardOffset()
TR::Compiler->target.cpu.maxUnconditionalBranchImmediateBackwardOffset()
```
Don't use macros at all (delete them).

0xdaryl commented on runtime/compiler/runtime/Trampoline.cpp:
```diff
@@ -858,6 +858,71 @@ void armCodeCacheParameters(int32_t *trampolineSize, void **callBacks, int32_t *
 
+#if defined(TR_TARGET_ARM64)
+
+#define TRAMPOLINE_SIZE         16 // ?
```
This is the number of bytes a single trampoline will occupy in the code cache.  It depends on the instructions you choose, and will likely be rounded to a multiple of 8.

0xdaryl replied:
It depends on the instructions you have in your trampoline.  If you're using the sequence I suggested below then I believe 16 should be sufficient (and its aligned on an 8-byte boundary too).

0xdaryl commented on runtime/compiler/runtime/Trampoline.cpp:
```diff
@@ -858,6 +858,71 @@ void armCodeCacheParameters(int32_t *trampolineSize, void **callBacks, int32_t *
+   }
+
+void arm64CreateHelperTrampolines(void *trampPtr, int32_t numHelpers)
+   {
```
The sequence you're after on AArch64 probably looks something like this:
```
   ldr r16, [PC+8]         ; r16 is the first intra-procedural scratch register
   br r16                   ; unconditional branch through register
   dq <address of target>  ; cache the address you want to jump to here
```
I believe the size of this trampoline is 16 bytes.  You can use a similar shape for the method trampolines.

hsoontie replied:
Thank you! Is the "dq" part done with the existing `*buffer = (intptrj_t)runtimeHelperValue((TR_RuntimeHelper)i)` line? I can't seem to find "dq" in the manual.

0xdaryl replied:
"dq" was my way of specifying 8 bytes of data (that's what it is in x86 assembly).  Its not a real machine instruction, but 8 bytes of memory where you store the address of the target.

0xdaryl commented on runtime/compiler/runtime/Trampoline.cpp:
```diff
@@ -858,6 +858,71 @@ void armCodeCacheParameters(int32_t *trampolineSize, void **callBacks, int32_t *
+      // LDR  PC, [PC, #-4]
+      // DCD  helperAddr
+      //
+      *buffer = 0xe51ff004; // What should this encoding be?
```
I don't know the encoding exactly, but it is pretty easy to figure out.  Start with the instruction sequence you want to lay down in the trampoline (e.g., the `ldr; br; dq` sequence above).  You can then look up the encoding for each instruction from the ARM Architecture Reference manual.  See section C6.2.34 for BR and C6.2.111 for LDR.  It should be straightforward to compute the encodings from the diagrams in those sections.

0xdaryl commented on runtime/compiler/runtime/Trampoline.cpp:
```diff
@@ -858,6 +858,71 @@ void armCodeCacheParameters(int32_t *trampolineSize, void **callBacks, int32_t *
+   TR_LinkageInfo *linkInfo = TR_LinkageInfo::get(startPC);
+   intptrj_t dispatcher = (intptrj_t)((uint8_t *)startPC + linkInfo->getReservedWord());
+
+   // LDR  PC, [PC, #-4]
```
This should look similar to the helper trampolines, except of course for how the address gets laid down.

hsoontie commented:
@0xdaryl Here is an update. I changed some things from my previous addition, and also copied over and modified some code for `arm64CodePatching` to get started with it.
I have a couple of points of confusion about `armCodePatching`, namely surrounding what the parameters being passed represent. Would you be able to tell me anything about those?
Thank you!

0xdaryl commented on runtime/compiler/runtime/Trampoline.cpp:
```diff
@@ -858,6 +858,108 @@ void armCodeCacheParameters(int32_t *trampolineSize, void **callBacks, int32_t *
+   *buffer = dispatcher;
+   }
+
+bool arm64CodePatching(void *callee, void *callSite, void *currentPC, void *currentTramp, void *newAddrOfCallee, void *extra)
```
* callee is the J9Method of the target method
* callSite is the address of the branch instruction in some method to the target
* currentPC is the address of the target method that you wish to call
* currentTramp is the address of the trampoline created to reach that method
* newAddrOfCallee is the new address of the target method (which can happen if the method is recompiled).  In such a case, the existing trampoline needs to be updated to branch to the new target.
* extra, not really sure what this is for.  Seems to be used only on Power to adjust some of the distance measurements.

hsoontie replied:
Thank you!

hsoontie commented:
After looking at your answer, there is a bit of code that is still confusing me at this point:

On line 897:
`uint8_t        *entryAddress = (uint8_t *)newAddrOfCallee + linkInfo->getReservedWord();`
I am just not sure what `entryAddress` is, because `getReservedWord` is not showing up in searches. The search function in my code editor has been giving me grief lately.

On lines 903 and 904:
```
distance = entryAddress - (uint8_t *)callSite - 8; // ?
currentDistance = (branchInstr << 8) >> 6; // ?
```
I don't understand why 8 is subtracted, or the purpose of the `>>` `<<` operations.

Any help with these would be appreciated!

hsoontie commented:
@0xdaryl , I forgot to tag you this time, sorry!

0xdaryl commented on runtime/compiler/runtime/Trampoline.cpp:
```diff
@@ -858,6 +858,108 @@ void armCodeCacheParameters(int32_t *trampolineSize, void **callBacks, int32_t *
+      }
+
+   if (TR::Options::getCmdLineOptions()->getOption(TR_StressTrampolines) 
+            || distance>(intptrj_t)TR::Compiler->target.cpu.maxUnconditionalBranchImmediateForwardOffset() 
```
Rather than these 2 target.cpu queries, I suggest you just use the `isTargetWithinUnconditionalBranchImmediateRange` API instead.

0xdaryl replied:
I think I had suggested once to use the two queries, but I think just the single, consolidated one is better (sorry!)

0xdaryl commented on runtime/compiler/runtime/Trampoline.cpp:
```diff
@@ -858,6 +858,108 @@ void armCodeCacheParameters(int32_t *trampolineSize, void **callBacks, int32_t *
+   int32_t         branchInstr = *(int32_t *)callSite;
+   void           *newTramp;
+
+   distance = entryAddress - (uint8_t *)callSite - 8; // ?
```
Where does the -8 come from?  I believe branch instruction displacements are relative to the start of the branch instruction, which I believe is simply `callSite`.

0xdaryl commented on runtime/compiler/runtime/Trampoline.cpp:
```diff
@@ -858,6 +858,108 @@ void armCodeCacheParameters(int32_t *trampolineSize, void **callBacks, int32_t *
+      else
+         {
+         newTramp = mcc_replaceTrampoline(reinterpret_cast<TR_OpaqueMethodBlock *>(callee), callSite, currentTramp, currentPC, newAddrOfCallee, false);
+         //TR_ASSERT(newTramp != NULL, "This is an internal error.\n");
```
Probably worth a FATAL assert if this happens.

0xdaryl commented:
Oops.  I didn't see your comments above.  I'll address them.

0xdaryl commented:
> I am just not sure what entryAddress is, because getReservedWord is not showing up in searches. 

The LinkageInfo word is a 32-bit structure that lives just before the interpreter entry point in Java compiled methods.  It contains some metadata about the compiled body that is easily accessible from the code.  You can find it in OMR:

https://github.com/eclipse/omr/blob/5e6e968a9d930209eaa790f02e578535a1625bb7/compiler/runtime/Runtime.hpp#L98

> `distance = entryAddress - (uint8_t *)callSite - 8; // ?`

I don't think the -8 adjustment is necessary since the branch displacement should be relative to the start of the branch instruction (i.e., `callSite`).

> `currentDistance = (branchInstr << 8) >> 6; // ?`

This code must have been taken from the 32-bit ARM implementation.  What it's doing is extracting the displacement from an encoded branch instruction.  On 32-bit ARM, the `BL` instruction is encoded with 8-bits in the mnemonic and 24-bits in the displacement (see Section F 5.1.25).  The code above first shifts left by 8 bits to get the displacement in the highest bits of the word.  The displacement encoded in the instruction is actually implicitly multiplied by 4 when used by the processor.  So, to get the true displacement encoded in the instruction it is actually shifted right by 6 bits (equivalent to shifting right by 8-bits and then shifting left again by 2 bits to multiply it by 4).  The right shift is arithmetic and will extend the sign as it shifts.

For AArch64, you need to mimic the behaviour for the AArch64 branch instruction using the sizes of its mnemonics and displacement fields.  Its encoding is different as the displacement is wider (26-bits).  See section C6.2.31.

hsoontie commented:
Thank you Daryl! Yes, I will make a new PR for this in openj9 once I've updated it.
